### PR TITLE
[NETBEANS-3990]: Auto completion for Sealed Types

### PR DESCRIPTION
**Pull request by Akshay-Gupta-Oracle**
Auto completion of keywords sealed, non-sealed, permits and classes,records names inside permits clause.

### CODE REVIEW

File: java/java.completion/src/org/netbeans/modules/java/completion/JavaCompletionTask.java [hunk 1]
```diff
@@ -22,6 +22,7 @@
 import com.sun.source.tree.*;
 import com.sun.source.tree.Tree.Kind;
 import com.sun.source.util.*;
+import com.sun.tools.javac.tree.JCTree.JCExpression;
```
Comment (lahodaj) [inline java/java.completion/src/org/netbeans/modules/java/completion/JavaCompletionTask.java:25]: Please no internal types inside language features (besides the language infrastructure, like java.source{.base}, etc.).

File: java/java.completion/test/unit/data/goldenfiles/org/netbeans/modules/java/completion/JavaCompletionTaskTest/1.8/CreateConstructorNonDefaultConstructor.pass [hunk 2]
```diff
@@ -25,9 +25,11 @@ int
 interface
 long
 native
+non-sealed
```
Comment (lahodaj) [inline java/java.completion/test/unit/data/goldenfiles/org/netbeans/modules/java/completion/JavaCompletionTaskTest/1.8/CreateConstructorNonDefaultConstructor.pass:28]: I think I would prefer to only have the keywords for an appropriate source level. The existing tests are probably not setting source level to 14, so I would incline to not add the keywords there. New tests may set source level to 14 and check the keywords are proposed by the code completion.

File: java/java.completion/src/org/netbeans/modules/java/completion/JavaCompletionTask.java [hunk 3]
```diff
@@ -780,6 +813,13 @@ private void insideClass(Env env) throws IOException {
                 controller.toPhase(Phase.ELEMENTS_RESOLVED);
                 env.addToExcludes(controller.getTrees().getElement(path));
                 addTypes(env, EnumSet.of(INTERFACE, ANNOTATION_TYPE), null);
+            } else if (last != null && last.token().text().toString().equals("permits")) {
```
Comment (junichi11) [inline java/java.completion/src/org/netbeans/modules/java/completion/JavaCompletionTask.java:816]: should use `TokenUtilities.textEquals()`.
Please reformat each your change. (Select your changes > Source > Format)

**Comment by neilcsmith-net** [comment]
For same reason as #2282 moving to 12.2.

**Comment by arvindaprameya** [comment]
Approved, thanks for doing this Akshay for  12.2.

**Comment by Akshay-Gupta-Oracle** [comment]
@lahodaj , @junichi11 Does this looks OK to merge??

**Review by lahodaj — APPROVED** [review]
Looks reasonable to me.

File: java/java.completion/src/org/netbeans/modules/java/completion/JavaCompletionTask.java [hunk 4]
```diff
@@ -792,8 +830,14 @@ private void insideClass(Env env) throws IOException {
...
                     env.addToExcludes(controller.getTrees().getElement(path));
                     addTypes(env, EnumSet.of(INTERFACE, ANNOTATION_TYPE), null);
+                } else if (isSealedSupported(env) && last != null && TokenUtilities.textEquals(last.token().text(),PERMITS_KEYWORD)) {
+                    controller.toPhase(Phase.ELEMENTS_RESOLVED);
+                    env.addToExcludes(controller.getTrees().getElement(path));
+                    addClassTypes(env, null);
                 } else {
                     addKeyword(env, IMPLEMENTS_KEYWORD, SPACE, false);
+                   if(isSealedSupported(env)) 
+                       addKeyword(env, PERMITS_KEYWORD, SPACE, false);
```
Comment (junichi11) [inline java/java.completion/src/org/netbeans/modules/java/completion/JavaCompletionTask.java:840]: Please format and use `{}`

File: java/java.completion/src/org/netbeans/modules/java/completion/JavaCompletionTask.java [hunk 5]
```diff
@@ -858,13 +905,22 @@ private void insideClass(Env env) throws IOException {
...
                         if (!tu.isEnum(cls)) {
                             addKeyword(env, EXTENDS_KEYWORD, SPACE, false);
+                            if (isSealedSupported(env)) {
+                                addKeyword(env, PERMITS_KEYWORD, SPACE, false);
+                            }
                         }
                         if (!tu.isInterface(cls)) {
                             addKeyword(env, IMPLEMENTS_KEYWORD, SPACE, false);
-                        }
+                         }
```
Comment (junichi11) [inline java/java.completion/src/org/netbeans/modules/java/completion/JavaCompletionTask.java:923]: Please revert.

File: java/java.completion/src/org/netbeans/modules/java/completion/JavaCompletionTask.java [hunk 6]
```diff
@@ -4779,9 +4850,13 @@ private void addClassModifiers(Env env, Set<Modifier> modifiers) {
...
             kws.add(PUBLIC_KEYWORD);
         }
-        if (!modifiers.contains(FINAL) && !modifiers.contains(ABSTRACT)) {
+        if (!modifiers.contains(FINAL) && !modifiers.contains(ABSTRACT) && !modifiers.contains(SEALED_KEYWORD) && !modifiers.contains(NON_SEALED_KEYWORD)) {
             kws.add(ABSTRACT_KEYWORD);
             kws.add(FINAL_KEYWORD);
+            if(isSealedSupported(env)){
+                kws.add(SEALED_KEYWORD);
+                kws.add(NON_SEALED_KEYWORD);
+            }
```
Comment (junichi11) [inline java/java.completion/src/org/netbeans/modules/java/completion/JavaCompletionTask.java:4859]: Please format.

File: java/java.completion/src/org/netbeans/modules/java/completion/JavaCompletionTask.java [hunk 7]
```diff
@@ -780,6 +812,12 @@ private void insideClass(Env env) throws IOException {
                 controller.toPhase(Phase.ELEMENTS_RESOLVED);
                 env.addToExcludes(controller.getTrees().getElement(path));
                 addTypes(env, EnumSet.of(INTERFACE, ANNOTATION_TYPE), null);
+            } else if (isSealedSupported(env) && last != null && TokenUtilities.textEquals(last.token().text(),PERMITS_KEYWORD)) {
+                controller.toPhase(Phase.ELEMENTS_RESOLVED);
+                env.addToExcludes(controller.getTrees().getElement(path));
+                addClassTypes(env, null);
+            } else if(isSealedSupported(env)){
+                addKeyword(env, PERMITS_KEYWORD, SPACE, false);
```
Comment (junichi11) [inline java/java.completion/src/org/netbeans/modules/java/completion/JavaCompletionTask.java:820]: Please format.

**Comment by Akshay-Gupta-Oracle** [comment]
I've addressed and all the comments regarding formatting @junichi11. Squashed as well. Waiting for merge.

**Pull request merged**